File: 10.py
#!/usr/bin/python3

import re
from dataclasses import dataclass, field
from typing import Union

@dataclass
class Output:
    id: int
    chips: list[int] = field(default_factory=list)

    def giveChip(self, chip: int) -> None:
        self.chips.append(chip)

@dataclass
class Bot:
    id: int
    chips: list[int] = field(default_factory=list)
    commandQueue: list[str] = field(default_factory=list)

    def giveChip(self, chip: int) -> None:
        self.chips.append(chip)
        self.chips.sort()
        if (self.ready() and self.chips[0] == 17 and self.chips[1] == 61):
            print("Bot", self.id, "has chips 17 and 61")
        if (self.ready()):
            cmd = self.getNextCommand()
            if (cmd != None):
                parse_command(cmd)

    def ready(self) -> bool:
        return True if len(self.chips) == 2 else False
    
    def getLowHight(self) -> list[int]:
        result = self.chips
        self.chips = []
        return result

    def addCommandToQueue(self, command: str) -> None:
        self.commandQueue.append(command)

    def getNextCommand(self) -> Union[str, None]:
        if self.ready() and len(self.commandQueue) > 0:
            return self.commandQueue.pop()
        else:
            return None

def parse_command(line: str, checkQueue:bool = True) -> None:
    # print(bot_dict)
    # print(output_dict)
    match = re.match(r'value (\d+) goes to bot (\d+)', line)
    if match:
        chip = int(match.group(1))
        id = int(match.group(2))
        if (id in bot_dict):
            bot_dict[id].giveChip(chip)
        else:
            bot = Bot(id)
            bot.giveChip(chip)
            bot_dict[id] = bot
        return
    match = re.match(r'bot (\d+) gives low to (\w+) (\d+) and high to (\w+) (\d+)', line)
    if match:
        giver_id = int(match.group(1))
        if (giver_id not in bot_dict.keys()):
            bot = Bot(giver_id)
            bot.addCommandToQueue(line)
            bot_dict[giver_id] = bot
        elif (bot_dict[giver_id].ready()):
            if (checkQueue):
                bot_dict[giver_id].addCommandToQueue(line)
                cmd = bot_dict[giver_id].getNextCommand()
                parse_command(cmd, False)
            else:
                low, high = bot_dict[int(giver_id)].getLowHight()
                if (match.group(2) == 'output'):
                    id = int(match.group(3))
                    if (id not in output_dict):
                        output_dict[id] = Output(id)
                    output_dict[id].giveChip(low)
                else: # bot
                    id = int(match.group(3))
                    if (id not in bot_dict):
                        bot_dict[id] = Bot(id)
                    bot_dict[id].giveChip(low)
                if (match.group(4) == 'output'):
                    id = int(match.group(5))
                    if (id not in output_dict):
                        output_dict[id] = Output(id)
                    output_dict[id].giveChip(high)
                else: # bot
                    id = int(match.group(5))
                    if (id not in bot_dict):
                        bot_dict[id] = Bot(id)
                    bot_dict[id].giveChip(high)
        else:
            bot_dict[int(match.group(1))].addCommandToQueue(line)
        return

with open('10.dat') as f:
    data = f.read()
instruction_lines = data.split('\n')
instruction_lines.pop()
bot_dict = dict()
output_dict = dict()

for line in instruction_lines:
    parse_command(line)

print(output_dict[0].chips[0] * output_dict[1].chips[0] * output_dict[2].chips[0])

File: 10.php
<?php

class Output {
    public int $id;
    public array $chips;

    function __construct($id, $chips = array()) {
        $this->id = $id;
        $this->chips = $chips;
    }

    function giveChip(int $chip): void
    {
        array_push($this->chips, $chip);
    }
}

class Bot
{
    public int $id;
    public array $chips;
    public array $commandQueue;

    function __construct($id, $chips = array(), $commandQueue = array())
    {
        $this->id = $id;
        $this->chips = $chips;
        $this->commandQueue = $commandQueue;
    }

    function giveChip(int $chip): void
    {
        array_push($this->chips, $chip);
        sort($this->chips);
        if ($this->ready() && $this->chips[0] == 17 && $this->chips[1] == 61) {
            echo "Bot ".$this->id." has chips 17 and 61\n";
        }
        if ($this->ready()) {
            $cmd = $this->getNextCommand();
            if ($cmd !== null) {
                parse_command($cmd);
            }
        }
    }

    function ready(): bool
    {
        if (count($this->chips) == 2) {
            return true;
        }
        return false;
    }
    
    function getLowHight(): array
    {
        $result = $this->chips;
        $this->chips = [];
        return $result;
    }

    function addCommandToQueue(string $command): void
    {
        array_push($this->commandQueue, $command);
    }

    function getNextCommand(): string | null
    {
        if ($this->ready() && count($this->commandQueue) > 0) {
            return array_pop($this->commandQueue);
        } else {
            return null;
        }
    }
}

function parse_command(string $line, bool $checkQueue = true): void
{
    global $bot_dict;
    global $output_dict;

    preg_match('/value (\d+) goes to bot (\d+)/', $line, $match);
    if ($match) {
        $chip = intval($match[1]);
        $id = intval($match[2]);
        if (in_array($id, array_keys($bot_dict))) {
            $bot_dict[$id]->giveChip($chip);
        } else {
            $bot = new Bot($id);
            $bot->giveChip($chip);
            $bot_dict[$id] = $bot;
        }
        return;
    }
    preg_match('/bot (\d+) gives low to (\w+) (\d+) and high to (\w+) (\d+)/', $line, $match);
    if ($match) {
        $giver_id = intval($match[1]);
        if (!in_array($giver_id, array_keys($bot_dict))) {
            $bot = new Bot($giver_id);
            $bot->addCommandToQueue($line);
            $bot_dict[$giver_id] = $bot;
        } else if ($bot_dict[$giver_id]->ready()) {
            if ($checkQueue) {
                $bot_dict[$giver_id]->addCommandToQueue($line);
                $cmd = $bot_dict[$giver_id]->getNextCommand();
                parse_command($cmd, false);
            } else {
                [$low, $high] = $bot_dict[$giver_id]->getLowHight();
                if (($match[2]) == 'output') {
                    $id = intval($match[3]);
                    if (!in_array($id, array_keys($output_dict))) {
                        $output_dict[$id] = new Output($id);
                    }
                    $output_dict[$id]->giveChip($low);
                } else { // bot
                    $id = intval($match[3]);
                    if (!in_array($id, array_keys($bot_dict))) {
                        $bot_dict[$id] = new Bot($id);
                    }
                    $bot_dict[$id]->giveChip($low);
                }
                if (($match[4]) == 'output') {
                    $id = intval($match[5]);
                    if (!in_array($id, array_keys($output_dict))) {
                        $output_dict[$id] = new Output($id);
                    }
                    $output_dict[$id]->giveChip($high);
                } else { // bot
                    $id = intval($match[5]);
                    if (!in_array($id, array_keys($bot_dict))) {
                        $bot_dict[$id] = new Bot($id);
                    }
                    $bot_dict[$id]->giveChip($high);
                }
            }
        } else {
            $bot_dict[intval($match[1])]->addCommandToQueue($line);
        }
        return;
    }
}

$file = "10.dat";
$fin = fopen($file, "r");
$data = fread($fin, filesize($file));
$instruction_lines = preg_split('/\n/', $data, -1, PREG_SPLIT_NO_EMPTY);

$bot_dict = array();
$output_dict = array();

foreach ($instruction_lines as $line) {
    parse_command($line);
}

echo $output_dict[0]->chips[0] * $output_dict[1]->chips[0] * $output_dict[2]->chips[0];

fclose($fin);

?>
